src/functions/BaselineCorrection.C

```diff
@@ -0,0 +1,247 @@
+  applyCorrection();
+
+  // try building a linear interpolation object
+  try
+  {
+    _linear_interp = libmesh_make_unique<LinearInterpolation>(_time, _adj_accel);
+  }
+  catch (std::domain_error & e)
+  {
+    mooseError("In BaselineCorrection ", _name, ": ", e.what());
+  }
+}
+
+Real
+BaselineCorrection::value(Real t, const Point & /*P*/) const
+{
+  return _scale_factor * _linear_interp->sample(t);
+}
+
+void
+BaselineCorrection::applyCorrection()
+{
+  // store a reference to final array index
+  unsigned int index_end = _accel.size() - 1;
+
+  // Compute unadjusted velocity and displacment time histories
+  Real dt;
+  std::vector<Real> vel, disp;
+  vel.push_back(0);
+  disp.push_back(0);
+  for (unsigned int i = 0; i < index_end; ++i)
+  {
+    dt = _time[i + 1] - _time[i];
+
+    vel.push_back(BaselineCorrectionUtils::newmarkGammaIntegrate(
+        _accel[i], _accel[i + 1], vel[i], _gamma, dt));
+    disp.push_back(BaselineCorrectionUtils::newmarkBetaIntegrate(
+        _accel[i], _accel[i + 1], vel[i], disp[i], _beta, dt));
+  }
+
+  // initialize polyfits and adjusted time history arrays as the nominal ones
+  DenseVector<Real> coeffs;
+  _adj_accel = _accel;
+  std::vector<Real> p_fit, adj_vel = vel, adj_disp = disp;
+
+  // adjust time histories with acceleration fit if desired
+  if (isParamValid("accel_fit_order"))
+  {
+    _order = getParam<unsigned int>("accel_fit_order");
+    coeffs = BaselineCorrectionUtils::getAccelerationFitCoeffs(
+        _order, _adj_accel, _time, index_end, _gamma);
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+      adj_vel[i] -= p_fit[1];
+      adj_disp[i] -= p_fit[2];
+    }
+  }
+
+  // adjust with velocity fit
+  if (isParamValid("vel_fit_order"))
+  {
+    _order = getParam<unsigned int>("vel_fit_order");
+    coeffs = BaselineCorrectionUtils::getVelocityFitCoeffs(
+        _order, _adj_accel, adj_vel, _time, index_end, _beta);
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+      adj_disp[i] -= p_fit[2];
+    }
+  }
+
+  // adjust with displacement fit
+  if (isParamValid("disp_fit_order"))
+  {
+    _order = getParam<unsigned int>("disp_fit_order");
+    coeffs = BaselineCorrectionUtils::getDisplacementFitCoeffs(_order, adj_disp, _time, index_end);
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+    }
+  }
+}
+
+void
+BaselineCorrection::buildFromFile()
+{
+  // Read data from CSV file
+  MooseUtils::DelimitedFileReader reader(getParam<FileName>("data_file"), &_communicator);
+  reader.read();
+
+  // Check if specific column headers were input
```

Comment:
Do you need to always have the column header names? Currently, as far as I know, when we input accelerations to MASTODON, we are not required to have column header names.

Reply:
No this is not required, but, here, I enforced the case where if one variable was specified a header (e.g., time), then the other variable (acceleration), must also be specified a column header. This is to not confused the `DelimitedFileReader` class. If no header is specified for either, then the default is to read the first column as time and the second column as acceleration.